File: tst/cpp/NindSearch/NindSearch.cpp
//
// C++ Implementation: NindIndex_testLecteur
//
// Description: un petit moteur de recherche sur un lexique, fichier inverse et fichier des index locaux existants.
//
// Author: jys <jy.sage@orange.fr>, (C) LATECON 2014
//
// Copyright: 2014-2015 LATECON. See LICENCE.md file that comes with this distribution
// This file is part of NIND (as "nouvelle indexation").
// NIND is free software: you can redistribute it and/or modify it under the terms of the 
// GNU Less General Public License (LGPL) as published by the Free Software Foundation, 
// (see <http://www.gnu.org/licenses/>), either version 3 of the License, or any later version.
// NIND is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without 
// even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU Less General Public License for more details.
////////////////////////////////////////////////////////////
//#include "NindLexicon/NindLexicon.h"
#include "NindIndex/NindLexiconIndex.h"
#include "NindIndex/NindTermIndex.h"
#include "NindIndex/NindLocalIndex.h"
#include "NindIndex/NindIndexTest.h"
#include "NindIndex/NindDate.h"
#include "NindExceptions.h"
#include <time.h>
#include <string>
#include <list>
#include <set>
#include <iostream>
#include <fstream>
using namespace latecon::nindex;
using namespace std;
////////////////////////////////////////////////////////////
static void displayHelp(char* arg0) {
    cout<<"© l'ATÉCON"<<endl;
    cout<<"Programme d'interrogation d'une base nind préalablement indexée"<<endl;
    cout<<"L'utilisateur indique le terme cherché."<<endl;
    cout<<"Ce peut être un terme simple ou un terme composé."<<endl;
    cout<<"Un terme composé a la forme dièsée (ex \"kyrielle#outil#informatique\")."<<endl;
    cout<<"L'utilisateur indique ensuite le document examiné"<<endl;
    cout<<"et la localisation du terme cherché est affichée."<<endl;

    cout<<"usage: "<<arg0<<" --help"<< endl;
    cout<<"       "<<arg0<<" <fichier lexique>"<<endl;
    cout<<"ex :   "<<arg0<<" FRE.lexiconindex"<<endl;
}
////////////////////////////////////////////////////////////
#define OFF "\33[m"
#define BLUE "\33[0;34m"
#define BOLD "\033[1;31m"

////////////////////////////////////////////////////////////
int main(int argc, char *argv[]) {
    setlocale( LC_ALL, "French" );
    if (argc<2) {displayHelp(argv[0]); return false;}
    const string lexiconFileName = argv[1];
    if (lexiconFileName == "--help") {displayHelp(argv[0]); return true;}

    try {
        //calcule les noms des fichiers lexique et inverse
        const size_t pos = lexiconFileName.find('.');
        const string termindexFileName = lexiconFileName.substr(0, pos) + ".termindex";
        const string localindexFileName = lexiconFileName.substr(0, pos) + ".localindex";
        //le lexique lecteur
        NindLexiconIndex nindLexicon(lexiconFileName, false);
        NindIndex::Identification identification;
        nindLexicon.getIdentification(identification);
        //le fichier inverse lecteur
        NindTermIndex nindTermIndex(termindexFileName, false, identification);
        //le fichier des index locaux
        NindLocalIndex nindLocalIndex(localindexFileName, false, identification);
        //la classe d'utilitaires
        NindIndexTest nindIndexTest;
        cout<<"identification : "<<identification.lexiconWordsNb<<" termes, "<<identification.lexiconTime;
        cout<<" ("<<NindDate::date(identification.lexiconTime)<<")"<<endl;
        
        while (true) {
            char str [80];
            cout<<endl<<BLUE<<"Entrez le terme à rechercher : "<<OFF;
            cin.getline(str, 80, '\n');
            const string word = string(str);
            if (word.empty()) break;
            //le terme
            list<string> componants;
            nindIndexTest.split(word, componants);
            //trouve son identifiant
            const unsigned int ident = nindLexicon.getWordId(componants);
            if (ident == 0) {
                cout<<"INCONNU"<<endl;
                continue;
            }
            //recupere l'index inverse pour ce terme
            list<NindTermIndex::TermCG> termDef;
            nindTermIndex.getTermDef(ident, termDef);
            //affiche les documents dans lesquels est indexe ce terme
            for (list<NindTermIndex::TermCG>::const_iterator it1 = termDef.begin(); 
                 it1 != termDef.end(); it1++) {
                const NindTermIndex::TermCG &termCG = (*it1);
                cout<<BOLD<<"["<<ident<<"] "<<nindIndexTest.getCgStr(termCG.cg)<<OFF,
                cout<<" "<<termCG.frequency<<" fois dans ";
                const list<NindTermIndex::Document> &documents = termCG.documents;
                for (list<NindTermIndex::Document>::const_iterator it2 = documents.begin(); 
                     it2 != documents.end(); it2++) {
                    const NindTermIndex::Document &doc = (*it2);
                    cout<<doc.ident<<"("<<doc.frequency<<") ";
                }
                cout<<endl;
            }
            cout<<BLUE<<"Entrez le n° de doc à afficher : "<<OFF;
            cin.getline(str, 80, '\n');
            const unsigned int noDoc = atoi(str);
            //recupere l'index local du doc             
            list<NindLocalIndex::Term> localDef;
            nindLocalIndex.getLocalDef(noDoc, localDef);
            cerr<<"localDef.size()="<<localDef.size()<<endl;
            for (list<NindLocalIndex::Term>::const_iterator it3 = localDef.begin();
                    it3 != localDef.end(); it3++) {
                const NindLocalIndex::Term &term = (*it3);
                if (term.term == ident) {
                    cout<<nindIndexTest.getCgStr(term.cg)<<"<";
                    const list<NindLocalIndex::Localisation> &localisation = term.localisation;
                    string sep = "";
                    for (list<NindLocalIndex::Localisation>::const_iterator it4 = localisation.begin();
                            it4 != localisation.end(); it4++) {
                        cout<<sep<<(*it4).position<<"("<<(*it4).length<<")";
                        sep = " ";
                    }
                    cout<<"> ";
                }
            }
            cout<<endl;
        }
    }
    catch (FileException &exc) {cerr<<"EXCEPTION :"<<exc.m_fileName<<" "<<exc.what()<<endl; return false;}
    catch (exception &exc) {cerr<<"EXCEPTION :"<<exc.what()<< endl; return false;}
    catch (...) {cerr<<"EXCEPTION unknown"<< endl; return false; }
}

File: src/py/Nind_search.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
__author__ = "jys"
__copyright__ = "Copyright (C) 2017 LATEJCON"
__license__ = "GNU LGPL"
__version__ = "2.0.1"
# Author: jys <jy.sage@orange.fr>, (C) LATEJCON 2017
# Copyright: 2014-2017 LATEJCON. See LICENCE.md file that comes with this distribution
# This file is part of NIND (as "nouvelle indexation").
# NIND is free software: you can redistribute it and/or modify it under the terms of the 
# GNU Less General Public License (LGPL) as published by the Free Software Foundation, 
# (see <http://www.gnu.org/licenses/>), either version 3 of the License, or any later version.
# NIND is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without 
# even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU Less General Public License for more details.
import sys
from os import getenv, path
from time import ctime
from builtins import input
import NindFile
from NindLexiconindex import NindLexiconindex
from NindTermindex import NindTermindex
from NindLocalindex import NindLocalindex

def usage():
    if getenv("PY") != None: script = sys.argv[0].replace(getenv("PY"), '$PY')
    else: script = sys.argv[0]
    print ("""© l'ATEJCON.
Programme d'interrogation d'une base nind préalablement indexée et
spécifiée par un de ses fichiers.
L'utilisateur indique le terme cherché, simple ou composé.
Un terme composé a la forme blanc-souligné (ex kyrielle_outil_informatique).
Dans un second temps, l'utilisateur indique ensuite le document examiné
et la localisation du terme cherché est affichée.
Le format des fichiers est défini dans le document LAT2017.JYS.470.

usage   : %s <fichier lexiconindex> <terme cherché>
exemple : %s FRE.lexiconindex syntagme_nominal
"""%(script, script))

OFF = "\033[m"
BLUE = "\033[0;34m"
RED = "\033[1;31m"

def main():
    if len(sys.argv) < 3 :
        usage()
        sys.exit()
    nindFileName = path.abspath(sys.argv[1])
    terme = sys.argv[2].strip()
    
    #calcul des noms de fichiers (remplace l'extension)
    nn = nindFileName.split('.')
    nindlexiconindexName = '.'.join(nn[:-1])+'.nindlexiconindex'
    nindtermindexName = '.'.join(nn[:-1])+'.nindtermindex'
    nindlocalindexName = '.'.join(nn[:-1])+'.nindlocalindex'
    
    #ouvre les classes
    nindLexiconindex = NindLexiconindex(nindlexiconindexName)
    nindTermindex = NindTermindex(nindtermindexName)
    nindLocalindex = NindLocalindex(nindlocalindexName)
    
    #1) verifie l'identification des fichiers
    (maxIdentifiant, dateHeure) = nindLexiconindex.donneIdentificationFichier()
    print ("max=%d dateheure=%d (%s)"%(maxIdentifiant, dateHeure, ctime(int(dateHeure))))
    (maxIdentifiant2, dateHeure2) = nindTermindex.donneIdentificationFichier()
    if maxIdentifiant2 != maxIdentifiant or dateHeure2 != dateHeure:
        print ("%s NON APAIRÉ : max=%d dateheure=%d (%s)"%(termindexFileName, maxIdentifiant2, dateHeure2, ctime(int(dateHeure2))))
    (maxIdentifiant2, dateHeure2) = nindLocalindex.donneIdentificationFichier()
    if maxIdentifiant2 != maxIdentifiant or dateHeure2 != dateHeure:
        print ("%s NON APAIRÉ : max=%d dateheure=%d (%s)"%(localindexFileName, maxIdentifiant2, dateHeure2, ctime(int(dateHeure2))))  
    
    #2) trouve l'identifiant du terme
    motsSimples = terme.split('_')
    motId = nindLexiconindex.donneIdentifiant(motsSimples)
    if motId == 0:
        print ('INCONNU')
        sys.exit()
        
    #3) trouve les utilisations du terme dans le fichier inverse et les affiche
    termesCGList = nindTermindex.donneListeTermesCG(motId)
    for (categorie, frequenceTerme, docs) in termesCGList:
        docsListe = []
        for (noDoc, frequenceDoc) in docs: docsListe.append('%d(%d)'%(noDoc, frequenceDoc))
        print ('%s[%s] %s%s %d fois dans %s'%(RED, motId, NindFile.catNb2Str(categorie), OFF, frequenceTerme, ' '.join(docsListe)))

    #4) choisit un doc et affiche ce qui concerne le terme
    noDocStr = input("%sno doc : %s"%(BLUE, OFF))
    if not noDocStr.isdigit(): sys.exit()
    noDoc = int(noDocStr)
    termList = nindLocalindex.donneListeTermes(noDoc)
    resultat = []
    for (noTerme, categorie, localisationsList) in termList:
        if noTerme != motId: continue
        locListe = []
        for (localisationAbsolue, longueur) in localisationsList: locListe.append('%d(%d)'%(localisationAbsolue, longueur))
        resultat.append('%s<%s>'%(NindFile.catNb2Str(categorie), ' '.join(locListe)))
    print (' '.join(resultat))
                                                                                  

if __name__ == '__main__':
        main()
